source/agora/consensus/protocol/Nominator.d

```diff
@@ -646,27 +646,38 @@ extern(D):
                     envelope.statement.pledges.confirm_.ballot.value, ex);
                 return;
             }
+            const validators = last_block.header.validators.count();
+            auto signed = last_block.header.validators.setCount;
+            if (last_block.header.height > 1 // Genesis block is not signed
+                && signed <= validators / 2) // Not more than half have signed previous block
+            {
+                log.info("Waiting for more than half to sign the last block, signed={}",
+                    last_block.header.validators.setCount);
+                long waited = 0;
+                // Wait till more than half have signed previous block
+                while (1)
```

Comment:
Instead of blocking and waiting for the signatures here, we probably should put it back to the `queued_envelopes` queue and let the `envelopeProcessTask` to handle it again.

Reply:
I was thinking about that too. We should probably also do the same for the missing `tx` case so that we do not throw away that envelope. It would also be nice if we could trigger the tasks here to fetch signatures (or missing `tx`s) so we don;t have to wait for the timer to fire.